### PR TITLE
[SYCL][AOT][NVPTX][AMD][New offloading model] Add AOT support for NVIDIA and AMD backend in the clang-linker-wrapper for SYCL offloading

### PR DESCRIPTION
**Pull request by asudarsa**
This PR adds third-party (nvptx, amd) AOT support for SYCL offloading in the clang-linker-wrapper tool.
Following list of changes are found;

1. Code inside clang-linker-wrapper has been refactored to get it ready for non-SPIRV backends for SYCL offloading. Main changes are inside linkAndWrapDeviceFiles. Explanations are added as PR comments.
2. A new clang-linker-wrapper option to pass additional device library files for NVPTX backend has been added
3. An additional check to emit error when device library files are unexpectedly empty has been added
4. New tests have been added to test the compilation flow for the new changes and also existing tests have been modified to react to the additional check correctly.

E2E tests and more extensive driver tests will be added in the next stage.

Thanks

### CODE REVIEW

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:11087]: For AMD, we do not pass any SYCL device libraries.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:771]: So far, this function and runAOTCompile function have been reading in from the SYCL table file and writing back to it. They have been modified to now work on individual inputs. The reading/writing of SYCL table file have been moved to a higher level of call hierarchy.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:817]: Please see comments to runLLVMToSPIRVTranslation for explanation

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:844]: Please see comments to runLLVMToSPIRVTranslation for explanation

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:878]: Please see comments to runLLVMToSPIRVTranslation for explanation

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1214]: Extra check added to make sure that device library files are available when expected.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:2043]: Design so far worked well for SYCL + SPIRV target. For non-spirv targets, we need to call clang instead of spirv-translator+AOTCompile on each of the split file output of sycl-post-link. Hence sycl-post-link and parsing of it table output has been moved here.
Initial validation has been performed with E2E tests on local third party hardware.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1379]: sycl-post-link has been moved up as it is required for SYCL offloading for non-spirv targets as well.

**Comment by bader** [comment]
> 1. Explanations are added as PR comments.

Why not **_code_** comments?

File: clang/test/Driver/Inputs/libsycl-complex-amd.new.o
Comment (bader) [inline clang/test/Driver/Inputs/libsycl-complex-amd.new.o:1]: Please, commit source files. Binary files are unmaintainable.

**Comment by asudarsa** [comment]
> > 1. Explanations are added as PR comments.
> 
> Why not **_code_** comments?

These comments are meant for reviewers who have been involved in the incremental development of this feature support. It explains what existed in the previous version and how things evolved in this version as I added AOT support for third party hardware. These comments do not make much sense for users/developers looking at the final version of the code. Hope that is agreeable.

Thanks

File: clang/test/Driver/Inputs/libsycl-complex-amd.new.o
Comment (asudarsa) [reply clang/test/Driver/Inputs/libsycl-complex-amd.new.o:1]: These tests are placeholder tests and will disappear as soon as we add end-to-end tests (within the next month).
Please let me know if that's agreeable.

Thanks

Comment (bader) [reply clang/test/Driver/Inputs/libsycl-complex-amd.new.o:1]: I don't think "placeholder tests" is a good idea in general. I suggest we drop "placeholder tests" and commit only final tests.

Regarding end-to-end tests: it's better to have standalone tests for corresponding component compiler driver or clang-linker-wrapper. Running and debugging compiler components with end-to-end tests is much more difficult as it requires additional dependencies like SYCL runtime library or specialized hardware.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (bader) [reply clang/lib/Driver/ToolChains/Clang.cpp:11087]: This comment seems useful for code readers not just PR reviewers.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:771]: It makes sense to extend the code comments before the function:

"Converts `File` file from LLVM bitcode to SPIR-V format using `llvm-spirv` tool. `Args` argument provides additional arguments for `llvm-spirv` tool."

Similar descriptions will be useful for all modified methods below.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1214]:
```suggestion
  // Make sure that device library files are available.
  if (ExtractedDeviceLibFiles.empty()) {
```

**Review by bader — COMMENTED** [review]
>These comments are meant for reviewers who have been involved in the incremental development of this feature support. It explains what existed in the previous version and how things evolved in this version as I added AOT support for third party hardware. These comments do not make much sense for users/developers looking at the final version of the code. Hope that is agreeable.

Some of these comments for reviewers is better to have in code.

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (asudarsa) [reply clang/lib/Driver/ToolChains/Clang.cpp:11087]: - Ah..you are right. Let me take a look at all these comments and move them into the code if it seems like a better fit.
- Thanks so much for looking closer.

Sincerely

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:771]: Will add the comments. Thanks.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1214]: Will add this to the upcoming commit. Thanks

File: clang/test/Driver/Inputs/libsycl-complex-amd.new.o
Comment (asudarsa) [reply clang/test/Driver/Inputs/libsycl-complex-amd.new.o:1]: I looked around some more into our codebase. Seems like I might be able to generate these .o files on the fly. Let me give it a shot.
Thanks

**Comment by mdtoguchi** [comment]
overall - looks OK to me.  Given there are still some changes needed to improve the testing area, I'll have @hchilama take another look when the bits are ready as she is covering for me while I'm out.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:771]: I have made a full parse of the file and added detailed comments for all SYCL specific functions.

Thanks

File: clang/test/Driver/Inputs/libsycl-complex-amd.new.o
Comment (asudarsa) [reply clang/test/Driver/Inputs/libsycl-complex-amd.new.o:1]: Removed all .o files.

Thanks for the feedback.

**Comment by asudarsa** [comment]
> overall - looks OK to me. Given there are still some changes needed to improve the testing area, I'll have @hchilama take another look when the bits are ready as she is covering for me while I'm out.

I have made changes to driver tests that I have added over time (mainly for new offloading model testing) to remove all .o files and am now generating them on the fly. @hchilama, can you please take a look when convenient?

Thanks

**Review by hchilama — APPROVED** [review]
LGTM - Driver

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td
Comment (frasercrmck) [inline clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:153]: Note sure we should have `nvptx`-specific options if we can avoid it. It makes it harder to update if anything changes in the future with supported SYCL targets, or with how we compile AMD.

Would it be at all possible to link this option up with `-Xsycl-target-backend`? Something like `-Xsycl-target-backend -sycl-device-libraries=...`? Then if there were multiple targets you could specify `-Xsycl-target-backend=nvptx64-nvidia-cuda -sycl-device-libraries=... -Xsycl-target-backend=amdgcn-amd-amdhsa -sycl-device-libraries=...`?

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (frasercrmck) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1199]: Could we specify all of this new documentation with doxygen, like the rest of LLVM/clang? Then we should be able to get properly rendered documentation.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (frasercrmck) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:918]: Do we really need to perform this step - can't we just `return runAOTCompileIntelGPU(InputFile, Args)` directly? We're just forwarding on a `Expected<StringRef>` from one function to another.

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:153]: Hi @frasercrmck 

Thanks so much for taking a look. One of the caveats in this PR is that this is not expected to be the final version to be upstreamed. This is expected to be an intermediate solution that is expected to functionally work. One of the key steps we intend t perform before upstreaming will be to refactor how we have introduced non-generic clang-linker-wrapper options. About this particular option, I feel that the we might be able to replace this option completety.

Hope it's agreeable from this PR's perspective.

Sincerely

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1199]: I am adding detailed code comments. Again, this is another step that we intend to perform as part of pre-upstreaming effort. Thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:918]: Good point. I will update. Thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (bader) [inline clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:506]: If we want to apply proper Doxygen annotations, we should add `\p` before the parameter name (or `\param` in extended form). See https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments for code examples:
```suggestion
/// \param Filename is the input file that could be a SPIR-V file.
/// \param Args encompasses all arguments required for linking and wrapping device
/// code and will be parsed to generate options required to be passed into the
/// llvm-spirv tool.
```

File: clang/test/Driver/Inputs/test-sycl.o
Comment (bader) [inline clang/test/Driver/Inputs/test-sycl.o:1]: 👍 
It's a pity that we have almost 1Mb of unnecessary binary files in the git history.

File: clang/test/Driver/sycl-post-link-options-win.cpp
Comment (bader) [inline clang/test/Driver/sycl-post-link-options-win.cpp:26]:
```suggestion
// RUN:   -sycl-device-libraries=%t.devicelib.o \
```
Do we need setting `-sycl-device-library-location` option?

File: clang/test/Driver/Inputs/test-sycl.o
Comment (asudarsa) [reply clang/test/Driver/Inputs/test-sycl.o:1]: Ah..you are right......sorry about that.....:-(

File: clang/test/Driver/linker-wrapper-sycl-win.cpp
Comment (asudarsa) [reply clang/test/Driver/linker-wrapper-sycl-win.cpp:5]: Yes. We require some .o file with a valid device image. Driver is more or less 'static'. But the linker-wrapper is dynamic and will skip all steps if no valid device image is found.

Thanks

File: clang/test/Driver/linker-wrapper-sycl-win.cpp
Comment (asudarsa) [reply clang/test/Driver/linker-wrapper-sycl-win.cpp:10]: I think you are correct. I tried it locally and it does work as expected. We do have to pass --offload-new-driver as well to make sure packager is called and not bundler.

File: clang/test/Driver/linker-wrapper-sycl-win.cpp
Comment (asudarsa) [reply clang/test/Driver/linker-wrapper-sycl-win.cpp:7]: I am replacing all such commands with the top-level %clang command. So, we can ignore this comment.

File: clang/test/Driver/linker-wrapper-sycl-win.cpp
Comment (asudarsa) [reply clang/test/Driver/linker-wrapper-sycl-win.cpp:16]: I am no longer requiring a .bc fil e as I have moved to top-level command. I will use 'touch' command for the .cpp file.

Thanks

File: clang/test/Driver/sycl-post-link-options-win.cpp
Comment (asudarsa) [reply clang/test/Driver/sycl-post-link-options-win.cpp:26]: I had a check in my linker-wrapper code to look for empty library location and throw an error if it is not set. I think I do  not need to do that. Removed it.

Thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:506]: I am hoping to wrap up this PR and do a 'global' comments update (for all my clang-linker-wrapper changes) as a follow-up step. Hope that's agreeable.

Thanks

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (asudarsa) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:2164]: Use case is this: If we have t1.bc (cuda offloading, x86_64 target) and t2.bc (openmp offloading, x86_64 target), then t1.bc and t2.bc will be linked together  and then following images will be created: t.cuda.bc and t.openmp.bc

If we had SYCL offloading for t1.bc, then we will link the two files using our approach to create t.sycl.bc using our approach and link the two files using community approach to create t.openmp.bc

I am not seeing any red flags here. But, I do not have realistic use cases for these. I can add them as part of my E2E testing.

Comment (sarnex) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:2164]: Ah ok, I didn't think we would want to link them together.

**Review by sarnex — APPROVED** [review]
LGTM as incremental change as long as feedback is tracked and addressed in follow-up PR

File: clang/test/Driver/linker-wrapper-sycl-win.cpp
Comment (bader) [reply clang/test/Driver/linker-wrapper-sycl-win.cpp:10]: Right. I suggest we call the driver. It will be just one simple command instead of multiple commands.

**Comment by asudarsa** [comment]
Hi @bader 

Thanks so very much for the detailed review. I have learnt a bunch about writing better tests here. 
I have addressed all the comments except the one related to 'doxygen' documentation which will be addressed in an upcoming patch.

Sincerely

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (bader) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:503]: It will be cleaner to fix annotations for the comments added in this PR to avoid dependency on future changes (i.e. make this change self-contained). I don't want this to block the merge, so I'll approve and leave the decision on how to proceed with this pull request to @asudarsa and the other reviewers.

**Review by bader — APPROVED** [review]

**Review by maksimsab — APPROVED** [review]
clang-linker-wrapper part LGTM.

File: clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td
Comment (frasercrmck) [reply clang/tools/clang-linker-wrapper/LinkerWrapperOpts.td:153]: Ah okay, I wasn't aware of that. That's fine, then. I just worry about the impact of changing flags once there's code out there that might be relying on them.

File: clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp
Comment (frasercrmck) [reply clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:1188]: Okay, no problem.

**Review by frasercrmck — APPROVED** [review]

**Review by Naghasan — COMMENTED** [review]
@asudarsa I'm in agreement with Fraser about his comments but obviously everyone is. Just had a small minor comment. Change looks good, thanks.

**Pull request merged**
